### PR TITLE
Add utility to down_sample node ids based on label

### PR DESCRIPTION
**Pull request by mkolodner-sc**
**Scope of work done**

- Introduces a utility which downsamples node ids based on labels. This will get called in the training/inference loop by the user if they want to downsample their node IDs.
- Adds tests for introduced utilities

Where is the documentation for this feature?: N/A

Did you add automated tests or write a test plan?

***Updated Changelog.md?*** NO

***Ready for code review?:*** NO

### CODE REVIEW

File: python/gigl/utils/down_sampler.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    if isinstance(dataset.node_labels, Mapping):
+        if node_type is None:
+            raise ValueError(
+                "node_type must be provided if the dataset is heterogeneous"
+            )
+        labels: Feature = dataset.node_labels[node_type]
+    else:
+        labels = dataset.node_labels
+
+    labels.lazy_init_with_ipc_handle()
```
[the remark this answers is not in the document]
Comment (mkolodner-sc) [reply python/gigl/utils/down_sampler.py:77]: Added comment here.

[the remark this answers is not in the document]
Comment (mkolodner-sc) [reply python/gigl/utils/down_sampler.py:83]: This is always the case and doesn't have anything to do with our codebase -- GLT does this under-the-hood whenever these fields are de-serialized and re-serialized to have better performance I presume. Generally, we'll need to do this when we are using these `Feature` fields after a dataset has been built. Note that this is a no-op if the fields have already been initialized [[1](https://github.com/alibaba/graphlearn-for-pytorch/blob/main/graphlearn_torch/python/data/feature.py#L243-L244)]

> I guess what I'm getting at is why don't we need this pattern with other parts of the dataset?

Many of the GLT operations on the dataset or features also end up calling this lazy init themselves [[2](https://github.com/alibaba/graphlearn-for-pytorch/blob/main/graphlearn_torch/python/data/feature.py#L147)]. The reason why we need to call this explicitly is because we need to extract the `id2index` tensor from the `Feature` object.

[the remark this answers is not in the document]
Comment (mkolodner-sc) [reply python/gigl/utils/down_sampler.py:83]: Yes, outside of tests users will never need to interact with these Feature fields directly, as GLT takes care of it under-the-hood. 

> I think it will be very surprising if they need to call some init_... first before they access feature tensors.

Yes, I was surprised too the first time I tried inspecting these feature tensors and found these fields to be None. In the majority of cases though, users will not need to interact with these features. At least for this PR, this is still something we must do to safely acesss the values of the GLT `Feature` object.

File: python/gigl/utils/down_sampler.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import torch
+from graphlearn_torch.data import Feature
+
+from gigl.common.logger import Logger
+from gigl.distributed.dist_dataset import DistDataset
+from gigl.src.common.types.graph_data import NodeType
+
+logger = Logger()
+
+
```
Comment (yliu2-sc) [inline python/gigl/utils/down_sampler.py:13]: better as an object class that handles downsampling, @kmontemayor2-sc for opinion as well

File: python/gigl/utils/down_sampler.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+def down_sample_node_ids_from_labels(
+    node_ids: torch.Tensor,
+    id2idx: torch.Tensor,
```
Comment (yliu2-sc) [inline python/gigl/utils/down_sampler.py]: this can benefit from more descirption naming, ex. `node_id_to_index`

File: python/gigl/utils/down_sampler.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    node_label_feats: torch.Tensor,
```
Comment (yliu2-sc) [inline python/gigl/utils/down_sampler.py]: just `node_labels`?

File: python/gigl/utils/down_sampler.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+from gigl.src.common.types.graph_data import NodeType
+
+logger = Logger()
+
+
+def down_sample_node_ids_from_dataset_labels(
+    dataset: DistDataset,
+    node_ids: torch.Tensor,
+    node_type: Optional[NodeType] = None,
+    label_filter_fn: Callable[[torch.Tensor], torch.Tensor] = lambda x: x >= 0,
```
Comment (yliu2-sc) [inline python/gigl/utils/down_sampler.py:18]: does this actually achieve downsampling on majority class to minority class?

File: python/gigl/utils/down_sampler.py (hunk 2 above)
Comment (kmontemayor2-sc) [reply python/gigl/utils/down_sampler.py:13]: I prefer functions to objects generally - unless we have an explicit need for state.

File: python/gigl/utils/down_sampler.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmontemayor2-sc) [reply python/gigl/utils/down_sampler.py]: Sorry I'm a little confused - is all your code pushed? below is the file as I see it rn. Did we intend to remove the other function here?

Are we relying on some part of `dist_dataset` now?


```
from collections.abc import Callable, Mapping
from typing import Optional

import torch
from graphlearn_torch.data import Feature

from gigl.common.logger import Logger
from gigl.distributed.dist_dataset import DistDataset
from gigl.src.common.types.graph_data import NodeType

logger = Logger()


def down_sample_node_ids_from_dataset_labels(
    dataset: DistDataset,
    node_ids: torch.Tensor,
    node_type: Optional[NodeType] = None,
    label_filter_fn: Callable[[torch.Tensor], torch.Tensor] = lambda x: x >= 0,
) -> torch.Tensor:
    """
    Down sample node ids using the provided dataset's labels. Downsampling means that
    we will sample a subset of the node ids to use for any downstream purpose. The node label values will be used to
    determine what gets downsampled, and can be customized with the `label_filter_fn` argument.
    By default, this will filter out negative label values from the node ids.

    Args:
        dataset (DistDataset): The dataset to down_sample.
        node_ids (torch.Tensor): The node ids to down_sample. Should be a 1D tensor of shape (N,).
        node_type (Optional[NodeType]): The node type to down_sample. If the dataset is heterogeneous, this must be provided.
        label_filter_fn (Callable[[torch.Tensor], torch.Tensor]): A callable that takes a tensor of
            labels of shape [N, 1] and returns a boolean mask of the same [N, 1] shape, indicating which labels should be included.
            By default, filters out negative values.
    Returns:
        torch.Tensor: The down_sampled node ids of shape [M,], where M is the number of down_sampled node ids.
            Only nodes with valid labels are included in the output.
    """

    if isinstance(dataset.node_labels, Mapping):
        if node_type is None:
            raise ValueError(
                "node_type must be provided if the dataset is heterogeneous"
            )
        labels: Feature = dataset.node_labels[node_type]
    else:
        labels = dataset.node_labels

    labels_to_include = label_filter_fn(labels[node_ids]).squeeze(-1)

    down_sampled_node_ids = node_ids[labels_to_include]

    return down_sampled_node_ids
```
